saml2/saml2-service-provider/src/main/java/org/springframework/security/saml2/provider/service/web/Saml2AuthenticationTokenConverter.java

```diff
@@ -74,7 +79,27 @@ public Saml2AuthenticationToken convert(HttpServletRequest request) {
 		}
 		byte[] b = samlDecode(saml2Response);
 		saml2Response = inflateIfRequired(request, b);
-		return new Saml2AuthenticationToken(relyingPartyRegistration, saml2Response);
+		AbstractSaml2AuthenticationRequest authenticationRequest = loadAuthenticationRequest(request);
+		return new Saml2AuthenticationToken(relyingPartyRegistration, saml2Response, authenticationRequest);
+	}
+
+	/**
+	 * Use the given {@link Saml2AuthenticationRequestRepository} to load authentication
+	 * request.
+	 * @param authenticationRequestRepository the
+	 * {@link Saml2AuthenticationRequestRepository} to use
+	 * @since 5.6
+	 */
+	public void setAuthenticationRequestRepository(
+			Saml2AuthenticationRequestRepository<AbstractSaml2AuthenticationRequest> authenticationRequestRepository) {
+		this.authenticationRequestRepository = authenticationRequestRepository;
+	}
+
+	private AbstractSaml2AuthenticationRequest loadAuthenticationRequest(HttpServletRequest request) {
+		if (this.authenticationRequestRepository == null) {
```

Comment:
Instead of this null check, you might consider following the Null Object pattern, like so:

```java
private Function<HttpServletRequest, AbstractSaml2AuthenticationRequest> loader = (request) -> null;

// ...

public void setAuthenticationRequestRepository(Saml2AuthenticationRequestRepository requests) {
    // ...
    this.loader = requests::loadAuthenticationRequest;
}
```

This improves readability by removing identation and conditionals.